<a href="https://colab.research.google.com/github/reesha-rsh/MLb4/blob/main/Homework/HW8.%20Bagging_Random%20Forest.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# 1. Use any binary classification dataset
2. Define validation strategy and use it for all next steps without changes
3. Train decision tree model and estimate performance on validation

*   Validation approach: We have a small amount of data so I will use a **K Fold** method
*   Metric: I plan to optimize **fbeta score** with beta 0.5 to give more weight for precision, thus minimizing false positives - reducing the prediction that a passenger survived when he actually did not.



In [37]:
from sklearn.model_selection import train_test_split
import pandas as pd
import numpy as np


from google.colab import drive
drive.mount('/content/drive')

train_full = pd.read_csv("/content/drive/MyDrive/MLb4/EDA Titanic/train.csv")
test_full = pd.read_csv("/content/drive/MyDrive/MLb4/EDA Titanic/test.csv")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [38]:
def generate(df,age_median,fare_median):
  useless_features = ['Name','Ticket','Cabin']
  data_cleaned = df
  data_cleaned = data_cleaned.drop(columns = useless_features)

  # generate binary values using get_dummies
  data_cleaned = pd.get_dummies(data_cleaned, columns=['Sex'],prefix=["Sex"])
  data_cleaned = pd.get_dummies(data_cleaned, columns=['Embarked'],prefix=["Embarked"])

  # Check for NaN values in the DataFrame
  nan_mask = data_cleaned.isnull()
  # Count the number of NaN values in each column
  nan_count_per_column = data_cleaned.isnull().sum()

  data_cleaned['Age'] = data_cleaned['Age'].fillna(age_median)
  data_cleaned['Fare'] = data_cleaned['Fare'].fillna(fare_median)

  return data_cleaned


In [39]:
features_columns = ['Pclass',	'Age',	'SibSp',	'Parch',	'Fare',	'Sex_female',	'Sex_male',	'Embarked_C',	'Embarked_Q',	'Embarked_S']

In [40]:
# get medians that will fill NaNs in generate func
age_median = train_full['Age'].median()
fare_median = train_full['Fare'].median()

In [41]:
train = generate(train_full,age_median=age_median,fare_median=fare_median)
train

PassengerId  Survived  Pclass   Age  SibSp  Parch     Fare  Sex_female  \
0              1         0       3  22.0      1      0   7.2500           0   
1              2         1       1  38.0      1      0  71.2833           1   
2              3         1       3  26.0      0      0   7.9250           1   
3              4         1       1  35.0      1      0  53.1000           1   
4              5         0       3  35.0      0      0   8.0500           0   
..           ...       ...     ...   ...    ...    ...      ...         ...   
886          887         0       2  27.0      0      0  13.0000           0   
887          888         1       1  19.0      0      0  30.0000           1   
888          889         0       3  28.0      1      2  23.4500           1   
889          890         1       1  26.0      0      0  30.0000           0   
890          891         0       3  32.0      0      0   7.7500           0   

     Sex_male  Embarked_C  Embarked_Q  Embarked_S  
0           1           0           0           1  
1           0           1           0           0  
2           0           0           0           1  
3           0           0           0           1  
4           1           0           0           1  
..        ...         ...         ...         ...  
886         1           0           0           1  
887         0           0           0           1  
888         0           0           0           1  
889         1           1           0           0  
890         1           0           1           0  

[891 rows x 12 columns]

In [42]:
titanic_features = train[features_columns]
titanic_label = train['Survived']

In [43]:
from sklearn.model_selection import StratifiedKFold, GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import make_scorer, fbeta_score, classification_report, accuracy_score
from sklearn import metrics


In [44]:
random_state = 42

In [45]:
# Initialize the StratifiedKFold object
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=random_state)

In [46]:
# Define the custom scoring function with the desired beta value
beta = 0.5
custom_scorer = make_scorer(fbeta_score, beta=beta)

In [47]:
from sklearn.model_selection import StratifiedKFold, GridSearchCV
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import make_scorer, fbeta_score



# Define the DecisionTreeClassifier model
classifier = DecisionTreeClassifier(random_state=random_state)

# Define the hyperparameter grid to search over
param_grid = {
    'criterion': ['gini'],
    'splitter': ['random'],
    'max_depth': [None, 3, 5, 7, 9, 11],
    'min_samples_leaf': [1,  3,  5,  7,  9,  11],
    'max_features': [ 1,  3,  5,  7,  9,  10],
    'class_weight': ['balanced']
}


# Initialize the GridSearchCV object with the DecisionTreeClassifier, hyperparameter grid, and custom scorer
grid_search = GridSearchCV(classifier, param_grid, scoring=custom_scorer, cv=skf)

# Perform grid search to find the best hyperparameters
grid_search.fit(titanic_features, titanic_label)

# Get the best hyperparameters and the corresponding model
best_params = grid_search.best_params_
best_model = grid_search.best_estimator_


# Print the best hyperparameters
print("Best Hyperparameters:", best_params)

# Optionally, you can evaluate the model on the full data using the F-beta score with beta=0.1
y_pred = best_model.predict(titanic_features)

fbeta = fbeta_score(titanic_label, y_pred, beta=beta)
accuracy = accuracy_score(titanic_label, y_pred)

print("F-beta Score (beta={}): {:.4f}".format(beta, fbeta))
print("Accuracy: {:.4f}".format(accuracy))
print(metrics.classification_report(titanic_label, y_pred))
print(metrics.confusion_matrix(titanic_label, y_pred))


Best Hyperparameters: {'class_weight': 'balanced', 'criterion': 'gini', 'max_depth': 5, 'max_features': 3, 'min_samples_leaf': 3, 'splitter': 'random'}
F-beta Score (beta=0.5): 0.7950
Accuracy: 0.8249
              precision    recall  f1-score   support

           0       0.82      0.91      0.87       549
           1       0.83      0.69      0.75       342

    accuracy                           0.82       891
   macro avg       0.83      0.80      0.81       891
weighted avg       0.83      0.82      0.82       891

[[500  49]
 [107 235]]


# 4. Train bagging model with decision tree as a base model and estimate performance on validation

In [48]:
from sklearn.ensemble import BaggingClassifier


In [49]:
parameters = {
    "max_features": [1,  3,  5,  7,  9,  10],
    "max_samples": [0.7, 0.8, 0.9],
    "estimator__max_depth": [None, 3, 5, 7, 9, 11],
    "estimator__min_samples_leaf": [1,  3,  5,  7,  9,  11],
    'estimator__criterion': ['gini'],
    'estimator__splitter': ['random'],
    'estimator__class_weight': ['balanced']
}

dt = DecisionTreeClassifier(random_state=random_state)
bg = BaggingClassifier(dt, random_state=random_state, n_estimators=25)

bag_grid_search = GridSearchCV(bg, parameters, scoring=custom_scorer, cv=skf)


# Perform grid search to find the best hyperparameters
bag_grid_search.fit(titanic_features, titanic_label)

# Get the best hyperparameters and the corresponding model
best_params = bag_grid_search.best_params_
best_model = bag_grid_search.best_estimator_


# Print the best hyperparameters
print("Best Hyperparameters:", best_params)

# Optionally, you can evaluate the model on the full data using the F-beta score with beta=0.1
y_pred = best_model.predict(titanic_features)

fbeta = fbeta_score(titanic_label, y_pred, beta=beta)
accuracy = accuracy_score(titanic_label, y_pred)

print("F-beta Score (beta={}): {:.4f}".format(beta, fbeta))
print("Accuracy: {:.4f}".format(accuracy))
print(metrics.classification_report(titanic_label, y_pred))
print(metrics.confusion_matrix(titanic_label, y_pred))


Best Hyperparameters: {'estimator__class_weight': 'balanced', 'estimator__criterion': 'gini', 'estimator__max_depth': None, 'estimator__min_samples_leaf': 1, 'estimator__splitter': 'random', 'max_features': 5, 'max_samples': 0.7}
F-beta Score (beta=0.5): 0.9102
Accuracy: 0.9147
              precision    recall  f1-score   support

           0       0.91      0.96      0.93       549
           1       0.93      0.84      0.88       342

    accuracy                           0.91       891
   macro avg       0.92      0.90      0.91       891
weighted avg       0.92      0.91      0.91       891

[[527  22]
 [ 54 288]]


# RANDOM FOREST

In [50]:
from sklearn.ensemble import RandomForestClassifier

In [51]:
parameters = {
    "max_features": [1,  3,  5,  7,  9,  10],
    "max_samples": [0.7, 0.8, 0.9],
    "max_depth": [None, 3, 5, 7, 9, 11],
    "min_samples_leaf": [1,  3,  5,  7,  9,  11],
    'criterion': ['gini'],
    'class_weight': ['balanced']
}

rfc = RandomForestClassifier(n_estimators=25, random_state=random_state, n_jobs=-1, oob_score=True)
forest_grid_search = GridSearchCV(rfc, parameters, scoring=custom_scorer, cv=skf)

# Perform grid search to find the best hyperparameters
forest_grid_search.fit(titanic_features, titanic_label)

# Get the best hyperparameters and the corresponding model
best_params = forest_grid_search.best_params_
best_model = forest_grid_search.best_estimator_


# Print the best hyperparameters
print("Best Hyperparameters:", best_params)

# Optionally, you can evaluate the model on the full data using the F-beta score with beta=0.1
y_pred = best_model.predict(titanic_features)

fbeta = fbeta_score(titanic_label, y_pred, beta=beta)
accuracy = accuracy_score(titanic_label, y_pred)

print("F-beta Score (beta={}): {:.4f}".format(beta, fbeta))
print("Accuracy: {:.4f}".format(accuracy))
print(metrics.classification_report(titanic_label, y_pred))
print(metrics.confusion_matrix(titanic_label, y_pred))


Best Hyperparameters: {'class_weight': 'balanced', 'criterion': 'gini', 'max_depth': 9, 'max_features': 7, 'max_samples': 0.7, 'min_samples_leaf': 1}
F-beta Score (beta=0.5): 0.9105
Accuracy: 0.9169
              precision    recall  f1-score   support

           0       0.91      0.96      0.93       549
           1       0.93      0.85      0.89       342

    accuracy                           0.92       891
   macro avg       0.92      0.90      0.91       891
weighted avg       0.92      0.92      0.92       891

[[526  23]
 [ 51 291]]



# 5. Write your own bagging implementation:
  <br>5.1. Define init for our CustomBaggingClassifier
  <br>5.2. Write fit as described in lecture: divide train data on n parts (`n_estimators` in CustomBaggingClassifier), train `base_estimator` on each part and save these models inside class
  <br>5.3. For predictions we should use all saved models and combine their predictions (as voting)


In [52]:
class CustomBaggingClassifier:
    def __init__(self, base_estimator, n_estimators, max_samples=1.0, max_features=1.0, random_state=None):
        self.base_estimator = base_estimator
        self.n_estimators = n_estimators
        self.max_samples = max_samples
        self.max_features = max_features
        self.random_state = random_state
        self.estimators = {}

    def bootstrap_sampling(self, X, y):
        # Implement bootstrap sampling to randomly select subsets of data
        num_samples = X.shape[0]
        if isinstance(self.max_samples, int):
            sample_size = self.max_samples
        elif isinstance(self.max_samples, float):
            sample_size = int(self.max_samples * num_samples)
        else:
            raise ValueError("Invalid type for max_samples. It must be int or float.")

        n_features = X.shape[1]
        if isinstance(self.max_features, int):
            max_features = self.max_features
        elif isinstance(self.max_features, float):
            max_features = max(1, int(self.max_features * n_features))
        else:
            raise ValueError("Invalid type for max_features. It must be int or float.")

        np.random.seed(self.random_state)
        sample_indices = np.random.choice(num_samples, size=sample_size, replace=True)
        sample_features = np.random.choice(n_features, size=max_features, replace=True)

        X_sampled = X.iloc[sample_indices,sample_features]
        y_sampled = y.iloc[sample_indices]
        return X_sampled, y_sampled, sample_features

    def fit(self, X, y):


        for _ in range(self.n_estimators):
            # Get a random subset of the data using bootstrap sampling
            X_sampled, y_sampled, sample_features = self.bootstrap_sampling(X, y)

            # Train a new base estimator and store it in the list
            estimator = self.base_estimator.fit(X_sampled, y_sampled)
            self.estimators[estimator] = sample_features


    def predict(self, X):
        # Make predictions using majority voting for classification
        predictions = np.zeros((X.shape[0], len(self.estimators)))

        for index, (estimator,sample_features ) in enumerate(self.estimators.items()):

            predictions[:, index] = estimator.predict(X.iloc[:,sample_features])

        # Take the majority vote to make the final prediction
        final_predictions = np.apply_along_axis(lambda x: np.bincount(x.astype('int')).argmax(), axis=1, arr=predictions)

        return final_predictions

    def get_params(self, deep=True):
        return {
            'base_estimator': self.base_estimator,
            'n_estimators': self.n_estimators,
            'max_samples': self.max_samples,
            'max_features': self.max_features,
            'random_state': self.random_state
        }

    def set_params(self, **params):
        for param, value in params.items():
            setattr(self, param, value)
        return self


In [56]:
parameters = {
    "max_features": [1,  3,  5,  7,  9,  10],
    "max_samples": [0.7, 0.8, 0.9],
    "base_estimator__max_depth": [None, 3, 5, 7, 9, 11],
    "base_estimator__min_samples_leaf": [1,  3,  5,  7,  9,  11],
    'base_estimator__criterion': ['gini'],
    'base_estimator__splitter': ['random'],
    'base_estimatorr__class_weight': ['balanced']
}


dt = DecisionTreeClassifier(random_state=random_state)
cbg = CustomBaggingClassifier(dt, random_state=random_state, n_estimators=100)

cbag_grid_search = GridSearchCV(cbg, parameters, scoring=custom_scorer, cv=skf)


# Perform grid search to find the best hyperparameters
cbag_grid_search.fit(titanic_features, titanic_label)

# Get the best hyperparameters and the corresponding model
best_params = cbag_grid_search.best_params_
best_model = cbag_grid_search.best_estimator_


# Print the best hyperparameters
print("Best Hyperparameters:", best_params)

# Optionally, you can evaluate the model on the full data using the F-beta score with beta=0.1
y_pred = best_model.predict(titanic_features)

fbeta = fbeta_score(titanic_label, y_pred, beta=beta)
accuracy = accuracy_score(titanic_label, y_pred)

print("F-beta Score (beta={}): {:.4f}".format(beta, fbeta))
print("Accuracy: {:.4f}".format(accuracy))
print(metrics.classification_report(titanic_label, y_pred))
print(metrics.confusion_matrix(titanic_label, y_pred))


Best Hyperparameters: {'base_estimator__criterion': 'gini', 'base_estimator__max_depth': None, 'base_estimator__min_samples_leaf': 1, 'base_estimator__splitter': 'random', 'base_estimatorr__class_weight': 'balanced', 'max_features': 5, 'max_samples': 0.9}
F-beta Score (beta=0.5): 0.7935
Accuracy: 0.8103
              precision    recall  f1-score   support

           0       0.79      0.94      0.86       549
           1       0.86      0.60      0.71       342

    accuracy                           0.81       891
   macro avg       0.83      0.77      0.78       891
weighted avg       0.82      0.81      0.80       891

[[516  33]
 [136 206]]


# 6. Compare performance of sklearn bagging model with your own implementation

**CustomBaggingClassifier**

Best Hyperparameters: {'base_estimator__criterion': 'gini', 'base_estimator__max_depth': None, 'base_estimator__min_samples_leaf': 1, 'base_estimator__splitter': 'random', 'base_estimatorr__class_weight': 'balanced', 'max_features': 5, 'max_samples': 0.9}
F-beta Score (beta=0.5): 0.7935
Accuracy: 0.8103
              precision    recall  f1-score   support

           0       0.79      0.94      0.86       549
           1       0.86      0.60      0.71       342

    accuracy                           0.81       891
   macro avg       0.83      0.77      0.78       891
weighted avg       0.82      0.81      0.80       891

[[516  33]
 [136 206]]

BaggingClassifier

Best Hyperparameters: {'estimator__class_weight': 'balanced', 'estimator__criterion': 'gini', 'estimator__max_depth': None, 'estimator__min_samples_leaf': 1, 'estimator__splitter': 'random', 'max_features': 5, 'max_samples': 0.7}
F-beta Score (beta=0.5): 0.9102
Accuracy: 0.9147
              precision    recall  f1-score   support

           0       0.91      0.96      0.93       549
           1       0.93      0.84      0.88       342

    accuracy                           0.91       891
   macro avg       0.92      0.90      0.91       891
weighted avg       0.92      0.91      0.91       891

[[527  22]
 [ 54 288]]

I have a really big gap between Custom and sklearn bagging classifiers. I don't understand what I am missing in my custom implementation